scripts/linters/js_ts_linter.py

```diff
@@ -0,0 +1,944 @@
+        """
+
+        # Select JS files which need to be checked.
+        files_to_check = self.all_filepaths
+        parsed_js_and_ts_files = dict()
+        if not files_to_check:
+            return parsed_js_and_ts_files
+        compiled_js_dir = tempfile.mkdtemp(
+            dir=os.getcwd(), prefix='tmpcompiledjs')
+        if not self.verbose_mode_enabled:
+            python_utils.PRINT('Validating and parsing JS and TS files ...')
+        for filepath in files_to_check:
+            if self.verbose_mode_enabled:
+                python_utils.PRINT(
+                    'Validating and parsing %s file ...' % filepath)
+            file_content = FILE_CACHE.read(filepath)
+
+            try:
+                # Use esprima to parse a JS or TS file.
+                parsed_js_and_ts_files[filepath] = esprima.parseScript(
+                    file_content, comment=True)
+            except Exception as e:
+                # Compile typescript file which has syntax not valid for JS
+                # file.
+                if filepath.endswith('.js'):
+                    shutil.rmtree(compiled_js_dir)
+                    raise Exception(e)
+                try:
+                    compiled_js_filepath = self._compile_ts_file(
+                        filepath, compiled_js_dir)
+                    file_content = FILE_CACHE.read(compiled_js_filepath)
+                    parsed_js_and_ts_files[filepath] = esprima.parseScript(
+                        file_content)
+                except Exception as e:
+                    shutil.rmtree(compiled_js_dir)
+                    raise Exception(e)
+
+        shutil.rmtree(compiled_js_dir)
+
+        return parsed_js_and_ts_files
+
+    def _get_expressions_from_parsed_script(self):
+        """This function returns the expressions in the script parsed using
+        js and ts files.
+
+        Returns:
+            dict. contains the expressions in the script parsed using js
+            and ts files.
+        """
+
+        parsed_expressions_in_files = collections.defaultdict(dict)
+        components_to_check = ['controller', 'directive', 'factory', 'filter']
+
+        for filepath, parsed_script in self.parsed_js_and_ts_files.items():
+            parsed_expressions_in_files[filepath] = collections.defaultdict(
+                list)
+            parsed_nodes = parsed_script.body
+            for parsed_node in parsed_nodes:
+                for component in components_to_check:
+                    expression = _get_expression_from_node_if_one_exists(
+                        parsed_node, [component])
+                    parsed_expressions_in_files[filepath][component].append(
+                        expression)
+
+        return parsed_expressions_in_files
+
+    def _compile_ts_file(self, filepath, dir_path):
+        """Compiles a typescript file and returns the path for compiled
+        js file.
+        """
+        allow_js = 'true'
```

Comment:
If these are constants, I think you can just inline all these TBH.

I would extract stuff into a variable if the variable name actually has meaning, but here we're just repeating all the stuff that's already in cmd anyway, so I don't think anything is gained.

Reply:
Done.